File: Pythoncode/8.字符串转换整数(atoi).py
class Solution:
    def myAtoi(self, str: str) -> int:
        INT_MIN = - 2**31
        INT_MAX = 2**31 - 1
        if not str.split():
            return 0
        s = str.split()[0]
        try:
            number = int(s)
        except ValueError:
            numberSet2 = set(["0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "-", "+"])
            numberSet = set(["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"])
            if s[0] not in numberSet2:
                number =  0
            else:
                i = 1
                for i in range(1, len(s)):
                    if s[i] not in numberSet:
                        break
                try:
                    number = int(s[:i])
                except ValueError:
                    number = 0

        if number < INT_MIN:
            return INT_MIN
        elif number > INT_MAX:
            return INT_MAX
        else:
            return number

File: Cppcode/8.字符串转换整数(atoi).cpp
#include <string>
#include <climits>
#include <unordered_set>

using namespace std;
class Solution {
public:
    int myAtoi(string str) {
        int res = 0;
        int np;
        unordered_set<char> numbers = {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9'};
        auto index = str.find_first_not_of(' ');
        if (index == string::npos || (numbers.find(str[index]) == numbers.end() && str[index] != '+' && str[index] != '-'))
        {
            return 0;
        }
        if (str[index] == '-')
        {
            np = -1;
            index++;
        }
        else if (str[index] == '+')
        {
            np = 1;
            index++;
        }
        else
        {
            np = 1;
        }
        while (index < str.size() && numbers.find(str[index]) != numbers.end())
        {
            int pop = stoi(string(1, str[index])) * np;
            if (res > INT_MAX/10 || (res == INT_MAX / 10 && pop > 7)) return INT_MAX;
            if (res < INT_MIN/10 || (res == INT_MIN / 10 && pop < -8)) return INT_MIN;
            res = res * 10 + pop;
            index++;
        }
        return res;
    }
};